File: dart/hamming/lib/hamming.dart
import 'package:collection/collection.dart';

class Hamming {
  int distance(String strand1, String strand2) {
    int hammingDistance = 0;
    if (strand1.length == 0 && strand1.length == strand2.length) {
      return 0;
    }
    if (strand1.length == 0 || strand2.length == 0) {
      throw new ArgumentError('no strand must be empty');
    }
    if (strand1.length != strand2.length) {
      throw new ArgumentError("left and right strands must be of equal length");
    }
    hammingDistance = new IterableZip([strand1.runes, strand2.runes])
        .where((l) => l.first != l.last)
        .length;
    return hammingDistance;
  }
}

File: cpp/hamming/hamming.cpp
#include "hamming.h"

#include <string>
using namespace std;

namespace hamming
{
    int compute(string strand1, string strand2)
    {
        if (strand1.length() != strand2.length()){
            throw domain_error("The strands must be of the same length.");
        }
        int hamming_distance = 0;
        for (int i = 0; i < strand1.length(); i++)
        {
            if (strand1[i] != strand2[i])
            {
                hamming_distance++;
            }
        }
        return hamming_distance;
    }
} // namespace hamming
